feat: Storefront Components (#15651)
shopware/platform: src/Storefront/Resources/app/storefront/src/component-system/README.md

# JavaScript Component System

## Introduction

> For Twig components that have to implement interactive funcationality via JavaScript, we introduce a corresponding JavaScript component system, which can be seen as the successor of the former JS plugin system. There are some parts which will seem familiar if you aleady know the plugin system, but some parts were changed and improved.

Reviewer: ```suggestion
For Twig components that have to implement interactive functionality via JavaScript, we introduce a corresponding JavaScript component system, which can be seen as the successor of the former JS plugin system. There are some parts which will seem familiar if you already know the plugin system, but some parts were changed and improved.
```

### Major differences between plugin and component system.

1. **Automatic initialization**  

>     If the component is implemented properly it will automatically be initialized on the corresponding elements. Even if the DOM tree changes and elements are added or removed, the component will automatically be initiallized on added elements or destroyed for removed elements. No more manual re-initialization of plugins that have to work in conjunction after dynamic DOM changes.

Reviewer: ```suggestion
    If the component is implemented properly it will automatically be initialized on the corresponding elements. Even if the DOM tree changes and elements are added or removed, the component will automatically be initialized on added elements or destroyed for removed elements. No more manual re-initialization of plugins that have to work in conjunction after dynamic DOM changes.
```

2. **No registration needed**  

>     The component system uses native ES module loading that does everyhting for you, if you follow the conventions. The script will automatically be loaded and initialized on corresponding elements just based on the component's name.

Reviewer: ```suggestion
    The component system uses native ES module loading that does everything for you, if you follow the conventions. The script will automatically be loaded and initialized on corresponding elements just based on the component's name.
```

3. **Better events instead of overrides**  

>     The current override technique of the plugin system was not reintroduced to the component system, as it showed some major flaws, as overrides could only happen once which can lead to conflicts between different Shopware exntensions. Instead there is a central event system which is easier to use and offers a more robust public interface. In addtion, it offers special interception events, for example, to manupilate request data before it is send.

Reviewer: ```suggestion
    The current override technique of the plugin system was not reintroduced to the component system, as it showed some major flaws, as overrides could only happen once which can lead to conflicts between different Shopware extensions. Instead there is a central event system which is easier to use and offers a more robust public interface. In addition, it offers special interception events, for example, to manipulate request data before it is sent.
```

4. **No imports**  

>     We decided to make everything related to the component system available via global scope. This means it is available at the `window` object level and can directly be used in plain JavaScript. No imports or bundling is necessary. You can still use the bundling as it is avialable today or use your own build processes if desired, but the component scripts target for plain JavaScript that don't need to be build in conjuction with our core files.

Reviewer: ```suggestion
    We decided to make everything related to the component system available via global scope. This means it is available at the `window` object level and can directly be used in plain JavaScript. No imports or bundling is necessary. You can still use the bundling as it is available today or use your own build processes if desired, but the component scripts target for plain JavaScript that don't need to be build in conjunction with our core files.
```

## Overview

The JavaScript component system consists of two main parts:

1. **Shopware** - A new global singleton which acts as a central entry point.
2. **ShopwareComponent** - The abstract base class for all components.

## Component Architecture

### Base Component Class

All components extend from the `ShopwareComponent` base class, which provides:

- **Automatic initialization** based on CSS or data attribute selectors.
- **Lifecycle management** with init, destroy, and update methods.
- **Option merging** from data attributes and constructor parameters.
- **Mutation observation** for reactive updates to attributes or child elements.

### Global Shopware Instance

The `Shopware` class acts as a singleton that:

- Manages component loading via ES modules.
- Handles automatic component discovery and initialization.
- Provides methods for component communication.

> - Provices a central event system for cross component communication.

Reviewer: ```suggestion
- Provides a central event system for cross component communication.
```

    foo = "bar",
    custom = true,
    jsOptions = {},
%}

<div data-component="MyComponentNamespace:MyComponent"
     data-component-options='{{ jsOptions|json_encode }}'>

    {# Some component logic ... #}
</div>
```

If you want to have an even more component-style approach, you can simply pass through the Twig component properties to your JavaScript component.

```twig
{# views/component/MyComponent.html.twig #}

{% props
    foo = "bar",
    custom = true,
%}

<div data-component="MyComponentNamespace:MyComponent"
     data-component-options='{{ this.props|json_encode }}'>

    {# Some component logic ... #}
</div>
```

## Component Communication

### Event System

To react to actions from other components, there is a new central event system available which can be accessed via the global `window.Shopware` singleton.

In your component you can emit events to inform others about an action and pass additional data via the event.

```javascript
// MyComponent.js

export default class MyComponent extends ShopwareComponent {

    // ...

    doSomething() {
        const message = 'Hello World!';

        window.Shopware.emit('MyComponent:DoSomething', message);
    }
}
```

> Other components can the subscribe to this event to react on that.

Reviewer: ```suggestion
Other components can then subscribe to this event to react on that.
```

```javascript
// SomeOtherComponent.js

export default class SomeOtherComponent extends ShopwareComponent {

    init() {
        window.Shopware.on('MyComponent:DoSomething', (message) => {
            this.el.innerText = message;
        });
    }
}
```

Of course, you can also register to events from anywhere else, also from outside of the component system. For example, if you just want to extend the logic of an existing component.

### Event Interception

> In addition to the normal asynchronous events, there is a seprate event type which expects a return value that gets further processed within the component. These events make it even easier to extend a components logic and offers a bunch of different use cases, like manipulating request data before it gets send. 

Reviewer: ```suggestion
In addition to the normal asynchronous events, there is a separate event type which expects a return value that gets further processed within the component. These events make it even easier to extend a components logic and offers a bunch of different use cases, like manipulating request data before it gets send. 
```

For example the BuyButton component offers an event `BuyButton:PreSubmit` which is interceptable, as it is called via `emitInterception()`. It is triggered when a user clicks the buy button of a product.

```javascript
// BuyButton.js

export default class BuyButton extends ShopwareComponent {

    // ...

    onFormSubmit(event) {
        event.preventDefault();

        let requestUrl = this.el.getAttribute('action');
        let formData = window.Shopware.serializeForm(this.el);

        ({ requestUrl, formData } = window.Shopware.emitInterception('BuyButton:PreSubmit', { requestUrl, formData }));

        window.Shopware.emit('BuyButton:Submit', requestUrl, formData);

>         window.Shopware.callPluginMethod('OffCanvasCart', 'openOffCanvas', requestUrl, formData);

Reviewer 2: Does this still work? I can only find `callPluginMethod` on `window.PluginManager` but not on `window.Shopware`.

Author: ✅ Fixed

    }
}
```

You can see that the event `BuyButton:PreSubmit` offers the opportunity to manipulate the `formData` before it gets sent. From any other script you can intercept this event and work with the arguments send via the event.

```javascript
// Intercept the buy button event
window.Shopware.intercept('BuyButton:PreSubmit', (data) => {

    data.formData.append('foo', 'bar');

    return data;
});
```

Don't forget to return the data again, so the component logic can work with it. 

> There can be multiple subscribers to a single event. They will all be executed in the order as they are registered. You can change the order by passing a priority parameter as an optional third option, when registering an event. By default all subscribers have the priority `0`. The higher the priority the earlier the subscriber is called in the chain. Also negative values are possible to move a subscriber further down the chain.

Reviewer: the prio system is analog to the Symfony one, right?

Author: I am not familiar with the Symfony one, to be honest. 😅

Reviewer: it is 👍 😁
